File: streampark-common/src/main/scala/org/apache/streampark/common/fs/FsOperator.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.streampark.common.fs

import org.apache.streampark.common.enums.StorageType

object FsOperator {

  lazy val lfs: FsOperator = FsOperator.of(StorageType.LFS)

  lazy val hdfs: FsOperator = FsOperator.of(StorageType.HDFS)

  def of(storageType: StorageType): FsOperator = {
    storageType match {
      case StorageType.HDFS => HdfsOperator
      case StorageType.LFS => LfsOperator
      case _ => throw new UnsupportedOperationException(s"Unsupported storageType:$storageType")
    }
  }

}

abstract class FsOperator {

  def exists(path: String): Boolean

  def mkdirs(path: String): Unit

  def mkdirsIfNotExists(path: String): Unit = {
    if (!exists(path)) {
      mkdirs(path)
    }
  }

  def delete(path: String): Unit

  def mkCleanDirs(path: String): Unit

  def upload(srcPath: String, dstPath: String): Unit =
    upload(srcPath, dstPath, delSrc = false, overwrite = true)

  def copy(srcPath: String, dstPath: String): Unit =
    copy(srcPath, dstPath, delSrc = false, overwrite = true)

  def copyDir(srcPath: String, dstPath: String): Unit =
    copyDir(srcPath, dstPath, delSrc = false, overwrite = true)

  def upload(
      srcPath: String,
      dstPath: String,
      delSrc: Boolean = false,
      overwrite: Boolean = true): Unit

  def copy(
      srcPath: String,
      dstPath: String,
      delSrc: Boolean = false,
      overwrite: Boolean = true): Unit

  def copyDir(
      srcPath: String,
      dstPath: String,
      delSrc: Boolean = false,
      overwrite: Boolean = true): Unit

  def move(srcPath: String, dstPath: String): Unit

  def fileMd5(path: String): String

}

File: streampark-common/src/main/java/org/apache/streampark/common/fs/FsOperator.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.common.fs;

import org.apache.streampark.common.enums.StorageType;

/** File system operator factory and shared operations. */
public abstract class FsOperator {

    /** Scala-friendly local file system operator singleton. */
    public static final FsOperator lfs = new FsOperator() {

        @Override
        public boolean exists(String path) {
            return LfsOperator.exists(path);
        }

        @Override
        public void mkdirs(String path) {
            LfsOperator.mkdirs(path);
        }

        @Override
        public void delete(String path) {
            LfsOperator.delete(path);
        }

        @Override
        public void mkCleanDirs(String path) {
            LfsOperator.mkCleanDirs(path);
        }

        @Override
        public void upload(String srcPath, String dstPath, boolean delSrc, boolean overwrite) {
            LfsOperator.upload(srcPath, dstPath, delSrc, overwrite);
        }

        @Override
        public void copy(String srcPath, String dstPath, boolean delSrc, boolean overwrite) {
            LfsOperator.copy(srcPath, dstPath, delSrc, overwrite);
        }

        @Override
        public void copyDir(String srcPath, String dstPath, boolean delSrc, boolean overwrite) {
            LfsOperator.copyDir(srcPath, dstPath, delSrc, overwrite);
        }

        @Override
        public void move(String srcPath, String dstPath) {
            LfsOperator.move(srcPath, dstPath);
        }

        @Override
        public String fileMd5(String path) {
            return LfsOperator.fileMd5(path);
        }
    };

    /** Scala-friendly HDFS operator singleton. */
    public static final FsOperator hdfs = new FsOperator() {

        @Override
        public boolean exists(String path) {
            return HdfsOperator.exists(path);
        }

        @Override
        public void mkdirs(String path) {
            HdfsOperator.mkdirs(path);
        }

        @Override
        public void delete(String path) {
            HdfsOperator.delete(path);
        }

        @Override
        public void mkCleanDirs(String path) {
            HdfsOperator.mkCleanDirs(path);
        }

        @Override
        public void upload(String srcPath, String dstPath, boolean delSrc, boolean overwrite) {
            HdfsOperator.upload(srcPath, dstPath, delSrc, overwrite);
        }

        @Override
        public void copy(String srcPath, String dstPath, boolean delSrc, boolean overwrite) {
            HdfsOperator.copy(srcPath, dstPath, delSrc, overwrite);
        }

        @Override
        public void copyDir(String srcPath, String dstPath, boolean delSrc, boolean overwrite) {
            HdfsOperator.copyDir(srcPath, dstPath, delSrc, overwrite);
        }

        @Override
        public void move(String srcPath, String dstPath) {
            HdfsOperator.move(srcPath, dstPath);
        }

        @Override
        public String fileMd5(String path) {
            return HdfsOperator.fileMd5(path);
        }
    };

    protected FsOperator() {
    }

    public static FsOperator lfs() {
        return lfs;
    }

    public static FsOperator hdfs() {
        return hdfs;
    }

    public static FsOperator of(StorageType storageType) {
        switch (storageType) {
            case HDFS:
                return hdfs;
            case LFS:
                return lfs;
            default:
                throw new UnsupportedOperationException("Unsupported storageType:" + storageType);
        }
    }

    public abstract boolean exists(String path);

    public abstract void mkdirs(String path);

    public void mkdirsIfNotExists(String path) {
        if (!exists(path)) {
            mkdirs(path);
        }
    }

    public abstract void delete(String path);

    public abstract void mkCleanDirs(String path);

    public void upload(String srcPath, String dstPath) {
        upload(srcPath, dstPath, false, true);
    }

    public void copy(String srcPath, String dstPath) {
        copy(srcPath, dstPath, false, true);
    }

    public void copyDir(String srcPath, String dstPath) {
        copyDir(srcPath, dstPath, false, true);
    }

    public abstract void upload(String srcPath, String dstPath, boolean delSrc, boolean overwrite);

    public abstract void copy(String srcPath, String dstPath, boolean delSrc, boolean overwrite);

    public abstract void copyDir(String srcPath, String dstPath, boolean delSrc, boolean overwrite);

    public abstract void move(String srcPath, String dstPath);

    public abstract String fileMd5(String path);
}
